### PR TITLE
e2e, Introduce e2e test: resolving a FQDN entry

### PR DESCRIPTION
**Pull request by AlonaKaplan**
The test is creating a VMI with one multus and no default interface and verifies that `nslookup` to the interface fqdn retrieves the correct IP.

The PR also contains -
- Enable deploying kubevirt and cnao on cluster-up.
- Changing the number of nodes in the ci cluster to 1 (it is enough for now).
- Updating the functest.sh to run the e2e test.

### CODE REVIEW

**Comment by oshoval** [comment]
About nslookup binary
I think we should do the following as part of `hack/functest.sh`

`[ "$CI" == "true" ] && dnf install bind-utils` (*)
```
if ! nslookup www.google.com > /dev/null 2>&1;then
  echo error
  exit 1
fi
```

So on both local and on CI, it will directly fail as first step if the binary doesn't exists
the (*) will make it work on CI for now of course, and once we add it to the bootstrap image we can remove it
It might take time until it is merged, and a new image + bump exists, so IMO it is more comfortable to use (*) first
and do it in the bootstrap in parallel
bootstrap file `images/bootstrap/Dockerfile`

File: hack/functest.sh
Comment (oshoval) [inline hack/functest.sh:1]: please keep the previous shebang, this is the one we use all over

plus remove plesase the `set -xe`

File: tests/vm_startup_test.go
Comment (oshoval) [inline tests/vm_startup_test.go:117]: the fact you have an interface doesn't mean it has an ip yet theoretically
guest-agent might just advertise it bit later
maybe the ptp is the reason we can trust it here specifically

maybe better to make it generic ?

File: tests/vm_generator.go
Comment (oshoval) [inline tests/vm_generator.go:60]: why do we need this lablel?

File: tests/vm_generator.go
Comment (oshoval) [inline tests/vm_generator.go:91]: Lets divide into functions please if possible
```
func addNoCloudDisk(spec *v1.VirtualMachineInstanceSpec) *v1.VirtualMachineInstanceSpec {
	return addNoCloudDiskWitUserData(spec, "#!/bin/sh\n\necho 'printed from cloud-init userdata'\n")
}
```

File: hack/functest.sh
Comment (oshoval) [inline hack/functest.sh:8]: we have twice `-v`, lets remove the deprecated one

File: tests/vm_startup_test.go
Comment (oshoval) [inline tests/vm_startup_test.go:42]: Can be done as BeforeAll / BeforeSuite with a defer to cleanup it instead of the AfterEach
On the other hand, it guarantee the leftovers are cleaned if any, but implicit

File: tests/vm_startup_test.go
Comment (oshoval) [inline tests/vm_startup_test.go:101]: please use a const for it

File: tests/vm_startup_test.go
Comment (oshoval) [inline tests/vm_startup_test.go:105]: please use a const for it

File: tests/vm_startup_test.go
Comment (oshoval) [inline tests/vm_startup_test.go:108]: please use a const for it

**Review by oshoval — COMMENTED** [review]
Nice thanks
First pass

Please update PR desc to
e2e, Introduce e2e test: resolving a FQDN entry
or something like that

typo on the last line of the PR desc (file name missing dot)

**Comment by oshoval** [comment]
We need to remove push or pull_request from the git actions
when you create a PR it runs both, maybe because you are the original repo creator or something
(it doesn't happen when PRs are created by other people)

File: tests/vm_generator.go
Comment (oshoval) [inline tests/vm_generator.go:21]: Please use 128Mi
64Mi is not enough
see please https://github.com/kubevirt/kubevirt/pull/5039

File: hack/functest.sh
Comment (oshoval) [inline hack/functest.sh:8]: please add deterministic seed of 0 (unless we do want tests to run on random order once we have more)
on kubevirt we use fixed seed

**Comment by AlonaKaplan** [comment]
> `[ "$CI" == "true" ] && dnf install bind-utils` (*)
> 
> ```
> if ! nslookup www.google.com > /dev/null 2>&1; then
>   echo error
>   exit 1
> fi
> ```

Thanks, updated the PR

File: tests/vm_generator.go
Comment (AlonaKaplan) [reply tests/vm_generator.go:21]: done

File: cluster/up.sh
Comment (oshoval) [reply cluster/up.sh:19]: There is another flag `KUBVIRT_WITH_CNAO_SKIP_CONFIG`
I will do it on a follow PR

CNAO image is pre pulled as part of the kubevirtci provision, so the cluster node already has the image,
it saves time to pull the image when the pod of cnao is created

File: tests/vm_generator.go
Comment (AlonaKaplan) [reply tests/vm_generator.go:60]: All kubevirt example vmis have this label. I"m not sure why/if it is indeed needed.

File: tests/vm_generator.go
Comment (AlonaKaplan) [reply tests/vm_generator.go:91]: why?

File: tests/vm_startup_test.go
Comment (AlonaKaplan) [reply tests/vm_startup_test.go:28]: There is no chance that external contributor or even future us will understand why "127.0.0.1" is the dnsIP without thos comment.

File: tests/vm_startup_test.go
Comment (AlonaKaplan) [reply tests/vm_startup_test.go:29]: Well, it's not merged yet so I cannot consider it:) #36 will have to fix the e2e if the e2e will be merged first.

File: tests/vm_startup_test.go
Comment (AlonaKaplan) [reply tests/vm_startup_test.go:42]: Since we have only one test, and in the future will maybe have few test. I find this approach cleaner and safer.

File: tests/vm_startup_test.go
Comment (AlonaKaplan) [reply tests/vm_startup_test.go:101]: done

File: cluster/up.sh
Comment (oshoval) [reply cluster/up.sh:19]: Draft https://github.com/kubevirt/kubesecondarydns/pull/42
(didn't test yet), anyhow it shouldn't block you as it is orthogonal and can be reviewed after this PR

File: tests/vm_startup_test.go
Comment (AlonaKaplan) [reply tests/vm_startup_test.go:105]: why? It is never reused.

File: tests/vm_startup_test.go
Comment (AlonaKaplan) [reply tests/vm_startup_test.go:108]: why? It is never reused.

File: tests/vm_startup_test.go
Comment (AlonaKaplan) [reply tests/vm_startup_test.go:117]: done

File: tests/vm_startup_test.go
Comment (AlonaKaplan) [reply tests/vm_startup_test.go:158]: what do you mean by go style?

File: tests/vm_startup_test.go
Comment (oshoval) [reply tests/vm_startup_test.go:28]: so maybe just something generic as this is the flow of kubevirtci  ?
or even just the 2nd PR can be enough ?
two PRs is way too much imo

File: tests/vm_startup_test.go
Comment (oshoval) [reply tests/vm_startup_test.go:158]: like the VMI
https://github.com/kubevirt/kubesecondarydns/pull/38/files#diff-738c6c49cffa4043869737c9b30a528211b64516acbf533de44098f593e8b4d1R17

a go struct that is filled with values

File: tests/vm_generator.go
Comment (oshoval) [reply tests/vm_generator.go:60]: lets remove it then please? we don't need it

File: tests/vm_generator.go
Comment (oshoval) [reply tests/vm_generator.go:91]: it is nicer to have functions that hide the details and you just need to read the name when you read top down

File: tests/vm_startup_test.go
Comment (oshoval) [reply tests/vm_startup_test.go:42]: It is an overkill imo, anyhow we can refactor it later, but imo we should

File: hack/functest.sh
Comment (AlonaKaplan) [reply hack/functest.sh]: which one of the is deprecated?

File: tests/vm_startup_test.go
Comment (oshoval) [reply tests/vm_startup_test.go:105]: it is nicer to have a const even for one time variable which declares a global name

File: tests/vm_startup_test.go
Comment (oshoval) [reply tests/vm_startup_test.go:108]: it is nicer to have a const even for one time variable which declares a global name

File: hack/functest.sh
Comment (oshoval) [reply hack/functest.sh]: I think the ginkgo one
best to try please the different ways

Comment (AlonaKaplan) [reply hack/functest.sh]: Currently we have just one test, so it is not relevant.
Anyway, I think it is healthier to run the tests in a random order.

**Comment by oshoval** [comment]
Need to show this to Brian please
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubesecondarydns/38/pull-kubesecondarydns-e2e-k8s/1599692018617946112

**Comment by oshoval** [comment]
/test pull-kubesecondarydns-e2e-k8s

i hope btw sudo is not needed, lets see

File: hack/functest.sh
Comment (oshoval) [inline hack/functest.sh:5]: `-y`

**Comment by oshoval** [comment]
Created PR for the bootstrap https://github.com/kubevirt/project-infra/pull/2501

File: tests/vm_startup_test.go
Comment (oshoval) [reply tests/vm_startup_test.go:158]: Right maybe better to declare it as a string on top and use it ? (like kubevirt's `newNetworkAttachmentDefinition`)

btw alternative indentations if you prefer
This one is what user-guide of kubevirt use
```
  config: '{
      "cniVersion": "0.3.1",
      "name": "mynet",
      "plugins": [
        {
          "type": "ptp",
          "ipam": {
            "type": "static",
            "addresses": [
                {
                  "address": "10.10.0.5/24",
                  "gateway": "10.10.0.254"
                }
            ]
          }
        }
      ]
    }'
```
This one is what yq pretty prints

```
'{ "cniVersion": "0.3.1", "name": "mynet", "plugins": [ { "type": "ptp", "ipam": { "type": "static", "addresses": [ { "address": "10.10.0.5/24", "gateway": "10.10.0.254" } ] } } ] }'
```

File: tests/client.go
Comment (oshoval) [inline tests/client.go:27]: please check the err here

**Review by oshoval — COMMENTED** [review]
do we still need pkg/controllers/controllers_suite_test.go
or it can be deleted ?

**Comment by oshoval** [comment]
Please fork the repo, it seems you are using the main repo branches
This is why it runs both pull_request and push git actions.

File: hack/functest.sh
Comment (oshoval) [inline hack/functest.sh:12]: i think we need to add no-color
else we see the control characters on the log

File: tests/client.go
Comment (oshoval) [reply tests/client.go:27]: but NewManager might err when GetConfigOrDie doesnt

Comment (AlonaKaplan) [reply tests/client.go:27]: done

File: tests/vm_startup_test.go
Comment (oshoval) [inline tests/vm_startup_test.go:29]: what about reading it from the CR and building accordingly the expected string?
this way it will work for:
* empty string
* D/S that fill the cluster-name (if this what we are going to do at the end)
* non empty string

File: tests/vm_generator.go
Comment (AlonaKaplan) [reply tests/vm_generator.go:91]: I find it more readable to see we are adding here a NoCloudDisk with UserData. I don't see the value in hiding this detail.

File: tests/vm_startup_test.go
Comment (AlonaKaplan) [reply tests/vm_startup_test.go:28]: I think those two PR explain the best how we rich the dns. I prefer to have it and to same future developer some time in the future.

Comment (oshoval) [reply tests/vm_startup_test.go:28]: I am personally against too much comments / links
if we can find something in between it would be great please

File: tests/vm_generator.go
Comment (oshoval) [reply tests/vm_generator.go:90]: the more important thing here is the why
without it it the cirros is taking more time to load, something about the ipv4 stack
i think otherwise but won't insist

**Comment by AlonaKaplan** [comment]
> do we still need pkg/controllers/controllers_suite_test.go or it can be deleted ?

We need it. It is for the controller unit tests.

File: hack/functest.sh
Comment (AlonaKaplan) [reply hack/functest.sh:12]: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubesecondarydns/38/pull-kubesecondarydns-e2e-k8s/1599451475841388544
Can you please point me what control characters do we see here?

File: tests/vm_startup_test.go
Comment (AlonaKaplan) [reply tests/vm_startup_test.go:29]: I think it is an overkill for the test. If we will ever need it on d/s, then we can change it.

Comment (oshoval) [reply tests/vm_startup_test.go:29]: I think that this test might even help us to test it D/S even manually
but ok we can just update upon need,
what about reading it from os.getenv that allows to config it easily when running D/S ?

File: hack/functest.sh
Comment (AlonaKaplan) [reply hack/functest.sh:12]: We don't have the no-color flag in kubevirt. And I find it nice to see the colors. I prefer to keep it.

Comment (oshoval) [reply hack/functest.sh:12]: automation/test.sh:ginko_params="--no-color --seed=42"

from kubevirt

File: tests/vm_startup_test.go
Comment (oshoval) [reply tests/vm_startup_test.go]: note that even now the PRs are outdated, because the was another PR which change kuevirtci default host port from 53 to 31111
this is the problem with comments

I think we can mention here the README.md, that once the PR of readme is merged, it explains a bit about the topology we are using on kubevirtci

File: tests/vm_startup_test.go
Comment (oshoval) [reply tests/vm_startup_test.go:158]: what about the fixed ident ?

**Comment by oshoval** [comment]
Please note this test would run also as part of CNAO
CNAO deploys KSD and calls `make create-nodeport`, and then vendor the repo and run the `make functest` target.
so if there is something that related to kubeconfig / namespace etc
it should be dynamic (about kubeconfig saw you already took care)

**Comment by AlonaKaplan** [comment]
> 

I use test namespace. It should be ok.

**Comment by oshoval** [comment]
Once https://github.com/kubevirt/project-infra/pull/2504 is merged we can remove the dnf install line please

File: tests/vm_generator.go
Comment (oshoval) [inline tests/vm_generator.go:70]: we use virtio in couple places in the same file. worth to add const please

File: tests/vm_startup_test.go
Comment (oshoval) [inline tests/vm_startup_test.go:53]: isn't there a way to just delete in one shot all vms from a given namespace using the client ?
(if so, same for nads etc)

Note that the Expect in the for means that if an error occur, it will skip the rest, 
i guess it is fine for now, maybe in the future we would prefer to collect errors but continue and assert the errors at the end

File: tests/vm_startup_test.go
Comment (oshoval) [inline tests/vm_startup_test.go:70]: we have just one nad, can we simplify this AfterEach please ?

File: tests/vm_startup_test.go
Comment (oshoval) [inline tests/vm_startup_test.go:60]: please drop this blank line
(same for line 80)

File: tests/vm_startup_test.go
Comment (oshoval) [inline tests/vm_startup_test.go:59]: maybe return vmiList.Items and use BeEmpty ?
(same for the rest)

File: tests/vm_startup_test.go
Comment (oshoval) [inline tests/vm_startup_test.go:91]: Better and nicer please to delete once, and wait that get shows it is gone, as done with the rest
not delete in a loop as done now

File: tests/vm_startup_test.go
Comment (oshoval) [inline tests/vm_startup_test.go:98]: can be done please as a BeforeEach of this It (in the same Context), as it is not part of the "It" but preparation,
makes the It smaller

File: tests/vm_generator.go
Comment (oshoval) [inline tests/vm_generator.go:89]: do we need maybe to add quay.io ? so it won't take from docker.io which we might stop updating ?

**Review by oshoval — COMMENTED** [review]
Thanks, few more small comments

about PR desc, nodes reducing - maybe worth to either explain why, or at least tell see  (specific) commit for more info

File: hack/functest.sh
Comment (oshoval) [inline hack/functest.sh:5]: Please remove this line as https://github.com/kubevirt/project-infra/pull/2504 is merged and includes
`bind-utils`

**Comment by AlonaKaplan** [comment]
> Thanks, few more small comments
> 
> about PR desc, nodes reducing - maybe worth to either explain why, or at least tell see (specific) commit for more info

The relevant [commit](https://github.com/kubevirt/kubesecondarydns/pull/38/commits/aaf75c4b352b68f0e8b0179460ef59cc4b92c5d4) message has an explanation. Do you think it is not enough?

File: tests/vm_generator.go
Comment (oshoval) [reply tests/vm_generator.go:86]: Still you need to select which one - this is the important part.
and imo no need to add it to the local, as it you are fetching it anyhow
but whatever you prefer

File: tests/vm_startup_test.go
Comment (oshoval) [reply tests/vm_startup_test.go:52]: It isn't correct the Delete Delete .. until gone
it is correct to Delete, Get Get until gone imo

File: tests/vm_startup_test.go
Comment (oshoval) [inline tests/vm_startup_test.go:24]: why we need this with the random suffix ?
even if we run on parallel, there will be a namespace per instance as kubevirt is doing

File: tests/vm_generator.go
Comment (oshoval) [inline tests/vm_generator.go:11]: nit: const VirtIO
(this is like we are doing it on kubevirt)

File: tests/vm_startup_test.go
Comment (oshoval) [inline tests/vm_startup_test.go:32]: nit: please drop the blank line

File: hack/functest.sh
Comment (AlonaKaplan) [reply hack/functest.sh]: done

File: tests/vm_generator.go
Comment (AlonaKaplan) [reply tests/vm_generator.go:11]: done

File: tests/vm_startup_test.go
Comment (AlonaKaplan) [reply tests/vm_startup_test.go:59]: done.

File: tests/vm_generator.go
Comment (AlonaKaplan) [reply tests/vm_generator.go:86]: changed to `quay.io/kubevirt/cirros-container-disk-demo:devel`although it doesn't really matter what image we are using. The devel one is maintained by kubevirt is should be enough for us.

File: tests/vm_startup_test.go
Comment (oshoval) [reply tests/vm_startup_test.go:24]: It is not a good practice imo, and i less like the way using different namespaces
I prefer that we delete VMI and not wait for them but not delete the NS as we are doing on kubevirt
note that on kubevirt there are some places where we do wait for all resources to gone

**Review by oshoval — APPROVED** [review]
Thanks

**Pull request merged**
